### PR TITLE
Added medianet adapter

### PR DESCRIPTION
**Pull request by garrix-fan**
Adds Media.Net Bidder Adapter.
Vendor ID: 142
Maintainer Email: prebid@media.net

### CODE REVIEW

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Test
+    public void makeBidsShouldReturnBannerBidIfBannerIsPresent() throws JsonProcessingException {
+        final HttpCall<BidRequest> httpCall = sampleHttpCall(
+                DUMMY_REQUEST,
+                mapper.writeValueAsString(sampleBidResponse(bidBuilder -> bidBuilder.impid("123"))));
+        final Result<List<BidderBid>> result = medianetBidder.makeBids(httpCall, null);
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .containsOnly(BidderBid.of(Bid.builder().impid("123").build(), banner, "USD"));
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:111]: minor. containsExactly

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+            .containsOnly(DUMMY_REQUEST);
+    }
+
+    @Test
+    public void makeBidsShouldReturnErrorIfResponseBodyCouldNotBeParsed() {
+        final HttpCall<BidRequest> httpCall = sampleHttpCall(DUMMY_REQUEST, "invalid response");
+        final Result<List<BidderBid>> result = medianetBidder.makeBids(httpCall, null);
+        assertThat(result.getErrors()).hasSize(1);
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Failed to decode: Unrecognized token");
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:80]: let's change to:
```
assertThat(result.getErrors()).hasSize(1)
                .allMatch(error -> error.getType() == BidderError.Type.bad_server_response
                        && error.getMessage().startsWith("Failed to decode: Unrecognized token"));
```

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Test
+    public void makeHttpRequestsShouldNotModifyIncomingRequest() {
+        final Result<List<HttpRequest<BidRequest>>> result;
+        result = medianetBidder.makeHttpRequests(DUMMY_REQUEST);
+
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+            .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+            .containsOnly(DUMMY_REQUEST);
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:72]: minor. containsExactly

**Review by And1sS — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/medianet/MedianetBidder.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    private static String getEndpoint(String originalEndpoint, String replacement) {
+        try {
+            replacement = URLEncoder.encode(replacement, StandardCharsets.UTF_8.displayName());
+        } catch (UnsupportedEncodingException e) {
+            logger.error("Unable to replace params in medianet endpoint");
+        }
+        return StringUtils.replace(originalEndpoint, EXTERNAL_URL_MACRO, replacement);
+    }
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/medianet/MedianetBidder.java:36]: As I see, replacement will be only one time, on bidder instance creation, so pls update `{{PREBID_SERVER_ENDPOINT}}` with your value, we don't need to do it in bidder, also lines 18 and 19 can be removed

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+import java.util.List;
+import java.util.function.Function;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.banner;
+
+public class MedianetBidderTest extends VertxTest {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:28]: Pls check code style for tests, in examples like `AdprimeBidderTest`

File: src/test/resources/org/prebid/server/it/openrtb2/medianet/test-auction-medianet-request.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  "imp": [
+    {
+      "id": "test-imp-id",
+      "banner": {
+        "w": 300,
+        "h": 250
+      },
+      "ext": {
+        "medianet": {
+          "refresh_count": 1,
```
Comment (snahornyi) [inline src/test/resources/org/prebid/server/it/openrtb2/medianet/test-auction-medianet-request.json:12]: `refresh_count` is it just random field? Or it should be described in medianet.json?

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/medianet/MedianetBidder.java (hunk 4 above)
Comment (snahornyi) [reply src/main/java/org/prebid/server/bidder/medianet/MedianetBidder.java:36]: Or if everything as it should be, update code to 
```java
private static String getEndpoint(String endpoint, String replacement) {

        return endpoint.replace(EXTERNAL_URL_MACRO, HttpUtil.encodeUrl(replacement));
    }
```

Comment (garrix-fan) [reply src/main/java/org/prebid/server/bidder/medianet/MedianetBidder.java]: We have passed the modified url in MedianetConfiguration. We are using HttpUtil now.

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java (hunk 5 above)
Comment (garrix-fan) [reply src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:28]: Followed given-when-then style test cases.

File: src/test/resources/org/prebid/server/it/openrtb2/medianet/test-auction-medianet-request.json (hunk 6 above)
Comment (garrix-fan) [reply src/test/resources/org/prebid/server/it/openrtb2/medianet/test-auction-medianet-request.json]: It was a random field. We have updated the required fields in the medianet.json.

File: src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        return new BidderConfigurationProperties();
+    }
+
+    @Bean
+    BidderDeps medianetBidderDeps() {
+        return BidderDepsAssembler.forBidder(BIDDER_NAME)
+                .withConfig(configProperties)
+                .usersyncerCreator(UsersyncerCreator.create(externalUrl))
+                .bidderCreator(
+                    config -> new MedianetBidder(config.getEndpoint(), mapper, externalUrl)
```
[the remark this answers is not in the document]
Comment (garrix-fan) [reply src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java]: We need to classify various prebid server host requests at our end, passing host as a query param will help us with that.

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        final Result<List<BidderBid>> result = medianetBidder.makeBids(httpCall, null);
+        assertThat(result.getErrors()).hasSize(1);
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Failed to decode: Unrecognized token");
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_server_response);
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeBidsShouldReturnEmptyListIfBidResponseIsNull() throws JsonProcessingException {
+        final HttpCall<BidRequest> httpCall;
```
[the remark this answers is not in the document]
Comment (garrix-fan) [reply src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:85]: All the test cases are formatted following this rule.

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java (hunk 3 above)
Comment (garrix-fan) [reply src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java]: @And1sS Changed

**Comment by garrix-fan** [comment]
Hey, @And1sS @snahornyi .
We have made all the suggested changes. You may resolve the comments.
Thank You,
Tushar Singh,
Media.Net

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.banner;
+
+public class MedianetBidderTest extends VertxTest {
+
+    private static final String ENDPOINT_URL = "https://test.media.net?src=external.prebidserver.com";
+    private static final BidRequest DUMMY_REQUEST = BidRequest.builder()
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java:31]: We prefer to use givenBidRequest method, you can check examples in another test classes

File: src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                .withConfig(configProperties)
+                .usersyncerCreator(UsersyncerCreator.create(externalUrl))
+                .bidderCreator(this::getMedianetBidder)
+                .assemble();
+    }
+
+    private MedianetBidder getMedianetBidder(BidderConfigurationProperties config) {
+        String configEndpoint = config.getEndpoint();
+        String encodedExternalUrl = HttpUtil.encodeUrl(externalUrl);
+        String bidderEndpoint = configEndpoint.replace(EXTERNAL_URL_MACRO, encodedExternalUrl);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java:57]: 56 line can be merged to 57 and also we prefer to declare variables as `final`

Comment (snahornyi) [reply src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java:57]: And even more better, pls create create method `resolveEndpoint` and then pass it to constructor, which will be in `medianetBidderDeps` method

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java (hunk 10 above)
Comment (snahornyi) [reply src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java:57]:
```java
private String resolveEndpoint(String configEndpoint) {
        return configEndpoint.replace(EXTERNAL_URL_MACRO, HttpUtil.encodeUrl(externalUrl));
    }
```

Comment (garrix-fan) [reply src/main/java/org/prebid/server/spring/config/bidder/MedianetConfiguration.java]: Added this method, and calling it in medianetBidderDeps()

File: src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java (hunk 9 above)
Comment (garrix-fan) [reply src/test/java/org/prebid/server/bidder/medianet/MedianetBidderTest.java]: Changed to 'given' nomenclature

**Review by rpanchyk — APPROVED** [review]

**Review by snahornyi — APPROVED** [review]
LGTM! Thanks)

**Review by And1sS — APPROVED** [review]

**Pull request merged**